### ISSUE TITLE
Reproducing BLEU/chrF results

### ISSUE THREAD

**Issue by j0ma** [turkic-interlingua/til-mt#5]
Hi again and thanks for your help earlier on #4 . 

I managed to get the data downloaded and some models trained. However, I'm a bit unclear about what settings were used to compute the BLEU and chrF metrics in Table 3 of [1]. 

Based on [`get_sacrebleu`](https://github.com/turkic-interlingua/til-mt/blob/master/bilingual_baselines/scripts/evaluate.py#L25), it seems like the default tokenizer inside `sacrebleu` is used, which I believe is `v13`. Is this correct?

Furthermore, I was wondering if the system outputs are detokenized prior to computing the evaluation metrics? 
FWIW, the raw data in the corpus seems to still be in tokenized format (newlines added by me for clarity):
```
[user@mypc en-tr]$ head -n1 en-tr.en

An observer may not see or otherwise detect the damage ; perhaps 
it is too diffuse to critically affect one organ or system , or the individual 
in question has more of a vital amino acid , protein , or enzyme necessary 
to bond with and neutralize a particular contaminant .
```
Given the above and what I see in [evaluate.py](https://github.com/turkic-interlingua/til-mt/blob/master/bilingual_baselines/scripts/evaluate.py), I'm assuming that the BLEU scorer is fed pre-tokenized sentences?

As for chrF, I'm unable to find anything in evaluate.py or elsewhere. Only a mention of `chrf` in some of the Jupyter notebooks, referring to sacrebleu. Should I be using sacrebleu's chrF implementation or the one available e.g. in [nltk](https://www.nltk.org/api/nltk.translate.chrf_score.html)?

Thanks a lot in advance, once again. Please let me know if I'm wrong about any of my guesses.

[1] [Evaluating Multiway Multilingual NMT in the Turkic Languages](https://arxiv.org/abs/2109.06262)

**Comment by mirzakhalov** [comment]
Hey there! Thanks for bringing up this clarity issue. All of the evaluations were done using sacrebleu and, I am mostly sure, v13 was used as default in both cases. I did specify during the scoring to detokenize before computing the results. 

As for the chrf, since some models were trained using joeynmt, we used the one from that provides out of the box. As for the rest, I used NLTK's chrf implementation from here: https://www.nltk.org/_modules/nltk/translate/chrf_score.html

Let me know which pairs you are looking at and I can try to validate those as well!

**Comment by j0ma** [comment]
No problem. BLEU evaluation is tricky, and it's well known that it's hard to do right (see e.g. [this issue from huggingface](https://github.com/huggingface/evaluate/issues/105) and [this blog post by Matthias Mueller from UZH](https://bricksdont.github.io/posts/2020/12/computing-and-reporting-bleu-scores/).

Right now I'm only working on EN-TR and EN-UZ. I trained a transformer model that pretty closely replicates the hyperparameters from your paper. 

For EN-TR, I get the following on the in-domain test set:
```
tr-BLEU-13a-mixed	35.78992
tr-BLEU-intl-mixed	36.73881
tr-BLEU-char-mixed	54.11462
tr-BLEU-spm-mixed	38.56968
tr-BLEU-none-mixed	35.14181
tr-BLEU-13a-lc	36.25529
tr-BLEU-intl-lc	37.19543
tr-BLEU-char-lc	54.72371
tr-BLEU-spm-lc	38.96968
tr-BLEU-none-lc	35.60636
tr-CHRF3-min1-max6	54.24469
tr-CHRF2-char6-word0	52.7575
```

The different metrics correspond to different tokenizers inside `sacrebleu` as well as turning `lowercase` on/off. 
The metric `tr-BLEU-13a-mixed` corresponds to what sacrebleu computes by default.

As for the CHRF metrics, `tr-CHRF3-min1-max6` corresponds to the [NLTK implementation](https://github.com/j0ma/mrl_nmt22/blob/master/mrl_nmt/scoring.py#L126). The other one is sacrebleu's default parameters.

For EN-UZ, my results are a bit different and lag behind what was in the paper:

```
uz-BLEU-13a-mixed	38.25268
uz-BLEU-intl-mixed	38.28955
uz-BLEU-char-mixed	62.27629
uz-BLEU-spm-mixed	44.09174
uz-BLEU-none-mixed	32.02542
uz-BLEU-13a-lc	38.66349
uz-BLEU-intl-lc	38.69713
uz-BLEU-char-lc	62.68773
uz-BLEU-spm-lc	44.49003
uz-BLEU-none-lc	32.47428
uz-CHRF3-min1-max6	59.81726
uz-CHRF2-char6-word0	57.49279
```

However if I just let the model train further and disable gradient accumulation (`--update-freq` in fairseq) I get:

```
uz-BLEU-13a-mixed	46.4752
uz-BLEU-intl-mixed	46.78207
uz-BLEU-char-mixed	67.18167
uz-BLEU-spm-mixed	52.24269
uz-BLEU-none-mixed	44.01867
uz-BLEU-13a-lc	46.92273
uz-BLEU-intl-lc	47.22303
uz-BLEU-char-lc	67.53826
uz-BLEU-spm-lc	52.68088
uz-BLEU-none-lc	44.47322
uz-CHRF3-min1-max6	65.01501
uz-CHRF2-char6-word0	63.7507
```

So, overall my results seem to be "in the right ballpark". But looking at the variation in BLEU scores depending on the tokenizer and casing I want to make sure I am comparing apples to apples and get the evaluation settings right. That way it's easier to know the discrepancies come from training. :)
